build_tools/benchmarks/definitions/common.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from enum import Enum
+from typing import List
+
+
+class ArchitectureType(Enum):
+  """Type of architecture."""
+  CPU = "cpu"
+  GPU = "gpu"
+
+
+@dataclass(frozen=True)
+class ArchitectureInfo(object):
+  """Architecture information."""
+  type: ArchitectureType
+  architecture: str
+  microarchitecture: str
+
+
+class DeviceArchitecture(Enum):
+  """Predefined architecture/microarchitecture."""
+  # x86_64 CPUs
+  X86_64_CASCADELAKE = ArchitectureInfo(ArchitectureType.CPU, "x86_64",
+                                        "cascadelake")
+
+  # ARM CPUs
+  ARMV8_2_A_GENERIC = ArchitectureInfo(ArchitectureType.CPU, "armv8.2-a",
+                                       "generic")
+
+  # RISC-V CPUs
+  RV64_GENERIC = ArchitectureInfo(ArchitectureType.CPU, "rv64", "generic")
+  RV32_GENERIC = ArchitectureInfo(ArchitectureType.CPU, "rv32", "generic")
+
+  # Mobile GPUs
+  VALHALL_GENERIC = ArchitectureInfo(ArchitectureType.GPU, "valhall", "generic")
+  ADRENO_GENERIC = ArchitectureInfo(ArchitectureType.GPU, "adreno", "generic")
+
+  # CUDA GPUs
+  CUDA_SM70 = ArchitectureInfo(ArchitectureType.GPU, "cuda", "sm_70")
+  CUDA_SM80 = ArchitectureInfo(ArchitectureType.GPU, "cuda", "sm_80")
+
+
+class DevicePlatform(Enum):
+  """OS platform and ABI."""
+  LINUX_GNU = "linux-gnu"
+  LINUX_ANDROID29 = "linux-android29"
+
+
+class ModelSourceType(Enum):
+  """Type of model source."""
+  # Exported MLIR file.
+  EXPORTED_MLIR = "exported_mlir"
+  # Exported TFLite model file.
+  EXPORTED_TFLITE = "exported_tflite"
+  # Exported SavedModel from Tensorflow.
+  EXPORTED_TF = "exported_tf"
+
+
+class InputDataFormat(Enum):
+  """Model input data format."""
+  RANDOM = "random"
+  NUMPY_NPY = "numpy_npy"
+
+
+@dataclass(frozen=True)
+class DeviceSpec(object):
+  """Benchmark device specification."""
+  id: str
+  # Device vendor name. E.g., Pixel-6.
+  vendor_name: str
+  architecture: DeviceArchitecture
+  platform: DevicePlatform
+  # Device-specific configurations. E.g., big-cores. Benchmark machines use
+  # these configs to set up the devices.
+  configs: List[str]
```

Comment:
The other things are all properties of the device. This is a property of how we run something on the device. Do we want to have them together like this?

Reply:
I have an idea about including these "software configurable" device properties in the device spec. On a heterogeneous processor, little cores and big cores can be seen as the different devices. This field allows us to define device specs like `Pixel-6_4-big-cores`, `Pixel-6_4-little-cores`, ... So we can use the same 4-threads run config to benchmark on different 4-cores specs.

Also when comparing with other ML frameworks, even they have different run configs to specify 4-threads parallelism, they can refer to the same device spec, so we know they are run on the same big/little cores setting.